### ISSUE TITLE
Use nashorn in java 8 instead of clj-v8?

### ISSUE THREAD

**Issue by magnars** [magnars/optimus#14]
Would be nice to avoid the trouble with binaries, and also get support on windows.

**Comment by arbscht** [comment]
Update since this has come up again and I had some time to spare:

I've split the JSR-223 changes out of my clj-v8 fork, into clj-jsr223-v8. https://github.com/arbscht/clj-jsr223-v8/

This new project will be released independently but can loosely track clj-v8 releases (should any ever appear). It adapts clj-v8's API into a discoverable/pluggable JSR-223 form. So I expect Optimus can depend on clj-jsr223-v8 instead -- that is, after appropriate refactoring to use `javax.script`.

Once Optimus uses `javax.script`, the choice of scripting engine can be made configurable as a new feature, and we can play with Nashorn etc.

To recap: my next steps are to release clj-jsr223-v8; then try to assemble a pull request with updated script engine invocation in Optimus (replacing the direct dependency on clj-v8); and eventually make the script engines configurable.

How does that sound?

**Comment by magnars** [comment]
That sounds brilliant!
fre. 16. okt. 2015 kl. 07.34 skrev Abhishek Reddy <notifications@github.com

**Comment by arbscht** [comment]
Some recent news on this:

- I've updated  [clj-jsr223-v8](https://github.com/arbscht/clj-jsr223-v8/), a wrapper around clj-v8, that extends the pluggable JSR-223 API.
- There are releases of [clj-jsr223-v8 on Clojars](https://clojars.org/clj-jsr223-v8)
- I have a recent [fork of Optimus](https://github.com/arbscht/optimus/tree/jsr223-v8) that depends on clj-jsr223-v8 via `javax.script.*` instead of directly on clj-v8, and it seems to work.
  - This Optimus can be configured using [environ](https://github.com/weavejester/environ) to switch between backend engines with a simple switch, e.g. `OPTIMUS_JS_ENGINE=nashorn lein repl` will give you a REPL where Optimus will use Nashorn on JDK8 (among other ways of using environ).
  - All Optimus tests pass for me.

I think it's about ready for a PR. What do you think @magnars?

**Comment by magnars** [comment]
Oh, excellent. This looks great. That PR would be very welcome indeed. :)

**Comment by atroche** [comment]
wow, looking forward to this

**Comment by piotr-yuxuan** [comment]
Any news about this PR?

**Comment by magnars** [comment]
That's a good question. You can see the story of the PR here: https://github.com/magnars/optimus/pull/52. @arbscht was the primus motor for this fix, and it seems he may have run out of steam - it turned out to be quite the task. If someone who uses Windows wants to figure out what remains to be done and take over, I'm all for it.

We need to use JSR-223 instead of just defaulting to Nashorn always, because Nashorn turned out to be an order of magnitude slower than v8 when I did my testing. Minifying bootstrap.css with clean-css on Nashorn took ~10 seconds. That's a couple years ago now, so things are hopefully better now.
